Add Tailscale support

MarekPikula opened the pull request:
This PR started as provider-specific Tailscale support and is now being morphed into a **generic external overlay pattern**.

## Goal
Keep `kube-hetzner` core provider-agnostic while still enabling Tailscale/ZeroTier/Cloudflare WARP (and others) through outer modules.

## Direction
1. No provider-specific logic in core module internals.
2. Use generic hooks for connectivity and bootstrap.
3. Keep provider integrations in `examples/` and external wrappers.

## Implemented on upstream branch (`mysticaltech:tailscale`)
- Added generic `node_connection_overrides` input (node name -> SSH host override).
- Wired SSH/provisioner host selection to support external overlay endpoints.
- Added external overlay guidance in README and docs.
- Added `examples/external-overlay-tailscale/README.md` showing the outer-module approach.
- Removed stale provider-specific leftovers from old branch state (`enable_tailscale`, tailscale outputs/docs entries).

## Note
This PR head currently points to fork branch `serenitycode-dev:tailscale` with maintainer edits disabled. The updated implementation is pushed to upstream branch `mysticaltech:tailscale` (latest commit `86071ea`).

MarekPikula commented on locals.tf:
```diff
@@ -280,7 +286,8 @@ locals {
 
   base_firewall_rules = concat(
-    var.firewall_ssh_source == null ? [] : [
+    var.firewall_ssh_source == null ||
+    var.enable_tailscale.restrict_firewall ? [] : [
```
I would say that this is self-explanatory.

MarekPikula commented on modules/host/main.tf:
```diff
@@ -118,6 +118,129 @@ resource "hcloud_server" "server" {
+    user           = "root"
+    private_key    = var.ssh_private_key
+    agent_identity = local.ssh_agent_identity
+    host           = local.server_ip_tail
+    port           = var.ssh_port
```
Yes, it's an intended behaviour. The comment exists at definition of `server_ip_tail`. In the context of `terraform_data.tailscale_setup_change` it is used to apply the Tailscale configuration on change, but indeed it's a noop in case of the `enable_tailscale.enable = false`. I'll disable the block in case Tailscale is disabled.

MarekPikula commented on README.md:
```diff
@@ -1082,6 +1082,111 @@ _Thanks for the tip @carolosf._
+     count     = module.kube-hetzner[0].tailscale_devices
+     device_id = data.tailscale_device.nodes[count.index].id
+     routes    = module.kube-hetzner.tailscale_devices[count.index].routes
+   }
```
Fixed

MarekPikula committed:
Export control plane load balancer private IP

This allows the IP to be used for assigning a domain outside the module
in cases where the public interface is disabled, for example, when
Tailscale is used to route traffic to the control plane load balancer.

Signed-off-by: Marek Pikuła <marek@serenitycode.dev>

mysticaltech commented:
That's very good stuff @MarekPikula, thanks! Please if you can fix the new merge conflicts it would be great. Then I will properly review along with @valkenburg-prevue-ch.

MarekPikula committed:
tailscale: Add support for running Tailscale

This patch enables user to seamlessly bootstrap Tailscale on the node
allowing for a secure remote access without exposing SSH on the public
interface.

Signed-off-by: Marek Pikuła <marek@serenitycode.dev>

MarekPikula committed:
tailscale: Add option to enable subnet router

This patch makes it possible to use the Tailscale subnet router feature
with control plane load balancer, which enables highly available access
to the private IP, making it possible to disable the public interface on
the LB without the need to set up an additional jump host.

Signed-off-by: Marek Pikuła <marek@serenitycode.dev>

MarekPikula committed:
tailscale: Add option to restrict firewall

This makes it easy to disable SSH and Kube API on the node's public
interface (overrides `firewall_kube_api_source` and
`firewall_ssh_source`).

Signed-off-by: Marek Pikuła <marek@serenitycode.dev>

mysticaltech commented:
@MarekPikula Thanks a lot for all those good contributions, glad to see that this one is ready too. 

If all is good, it will be merged by the end of the week along with some other PRs, either as part v2.18 or v3.0.

alexandradeas commented:
Could this be updated to make the change less specific to tailscale? At some point in the future I'd like to run ZeroTier on each node, to do so it'd be a similar process of creating key pair, passing them into the node, and running a script to join the network. I would assume there's a similar process for CF Warp as well. Having a more generalised process of doing this by passing in something like a cloud-init script to run on each node would mean the module doesn't need to be updated to support more providers

Ducktatorrr commented:
Good stuff! I noticed you're checking whether [`[var.enable_tailscale.auth_key](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/1760/files#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288eR169)`](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/1760/files#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288eR169) starts with `tskey-`. Could this check be removed or updated with a comment noting that this prefix is specific to Tailscale? [[Headscale](https://headscale.net/)](https://headscale.net/), the open-source tailscale control server alternative, doesn't use the `tskey-` prefix for its preauth keys, so this might cause issues/confusion for users running Headscale.

mysticaltech commented:
That's the way @maaft! If anyone is inspired, please do shoot a PR, or you @MarekPikula please do consider something more generic that could work in your case with Tailscale too. 

If anyone is going to work on this, please let me know so I do not attempt it.

@MarekPikula I really appreciate this PR, but the all the reasons above, I will close it for now, the genericity is important, otherwise there are just too many external systems to support.

mikeywuu commented:
Just wanted to also let you know that I'm currently implementing something similar, achieving basically the same goals, by using Cloudflare(d) with Cloudflare's Zero Trust approach. The client in this case would be Cloudflare WARP.

I'm happy to also contribute this, maybe we can introduce a dedicated section for this since we would then already have three remote access solutions I think, being jump / bastion hosts with no public IPs on control/worker nodes; tailscale and cloudflare.

mysticaltech commented:
@alexandradeas Please do, I'm way over my head on time these days.

mysticaltech commented:
Superseded by #2055.

Closing this to continue on the new PR that preserves the original contribution lineage while applying the generic external-overlay approach discussed in this thread.
